libraries/stdlib/src/kotlin/collections/Grouping.kt

```diff
@@ -0,0 +1,278 @@
+): Map<K, S> =
+        aggregate { key, value, e, first ->
+            if (first) e else operation(key, value as S, e)
+        }
+
+/**
+ * Groups elements from the [Grouping] source by key and accumulates elements of each group with the specified [operation]
+ * starting the first element in that group to the given [destination] map.
+ *
+ * If the [destination] map already has a value corresponding to the key of some group,
+ * that value is used as an initial value of the accumulator for that group and the first element of that group is also
+ * subjected to the [operation].
+
+ * @param operation a function that is invoked on each subsequent element of the group with the following parameters:
+ *  - `accumulator`: the current value of the accumulator of the group;
+ *  - `element`: the element from the source being folded;
+ *
+ * @return the [destination] map associating the key of each group with the result of accumulating the group elements.
+ */
+@SinceKotlin("1.1")
+public inline fun <S, T : S, K, M : MutableMap<in K, S>> Grouping<T, K>.reduceTo(
+        destination: M,
+        operation: (key: K, accumulator: S, element: T) -> S
+): M =
+        aggregateTo(destination) { key, value, e, first ->
+            if (first) e else operation(key, value as S, e)
+        }
+
+
+/**
+ * Groups elements from the [Grouping] source by key and counts elements in each group.
+ *
+ * @return a [Map] associating the key of each group with the count of element in the group.
+ */
+@SinceKotlin("1.1")
+@JvmVersion
+public fun <T, K> Grouping<T, K>.eachCount(): Map<K, Int> =
+        // fold(0) { acc, e -> acc + 1 } optimized for boxing
+        fold(
+                initialValueSelector = { k, e -> kotlin.jvm.internal.Ref.IntRef() },
+                operation = { k, acc, e -> acc.apply { element += 1 } })
+        .mapValues { it.value.element }
+
+/**
+ * Groups elements from the [Grouping] source by key and counts elements in each group to the given [destination] map.
+ *
+ * @return the [destination] map associating the key of each group with the count of element in the group.
+ */
+@SinceKotlin("1.1")
+public fun <T, K, M : MutableMap<in K, Int>> Grouping<T, K>.eachCountTo(destination: M): M =
+        foldTo(destination, 0) { acc, e -> acc + 1 }
+
+/**
+ * Groups elements from the [Grouping] source by key and sums values provided by the [valueSelector] function for elements in each group.
+ *
+ * @return a [Map] associating the key of each group with the count of element in the group.
+ */
+@SinceKotlin("1.1")
+@JvmVersion
+public inline fun <T, K> Grouping<T, K>.eachSumOf(valueSelector: (T) -> Int): Map<K, Int> =
+        // fold(0) { acc, e -> acc + valueSelector(e)} optimized for boxing
+        fold(
+                initialValueSelector = { k, e -> kotlin.jvm.internal.Ref.IntRef() },
+                operation = { k, acc, e -> acc.apply { element += valueSelector(e) } })
+        .mapValues { it.value.element }
```

Comment:
I wouldn't call it internal API hack, because that API is used publicly by the compiler in the generated bytecode.
`mapInPlace` is possible as JVM-only optimization, otherwise it would make the code non-portable to platforms, where generics are reified.